Core api reference refresh (#669)
ROCm/rocDecode: docs/reference/rocDecode-parser.rst

.. meta::
  :description: The rocDecode parser API
  :keywords: parse video, parser, decode, video decoder, video decoding, rocDecode, core APIs, AMD, ROCm

********************************************************************

> The rocDecode parser API

Reviewer: ```suggestion
rocDecode parser API

```

Author: I like The The.

********************************************************************

> The rocDecode parser API, exposed in |rocparser|_, is used to decode bitstreams and organize them in a structured format that can be consumed by the hardware decoder.

Reviewer: ```suggestion
The rocDecode parser API, exposed in |rocparser|_, is used to decode bitstreams and organize them in a structured format that the hardware decoder can consume.

```

Author: It's a subtlety, but can be consumed has a different feel than can consume. The decoder isn't the object; the bitstreams are the object. They're being consumed.

Parsing parameters are stored in the ``RocdecParserParams`` struct and passed to ``rocDecCreateVideoParser()`` to create a new parser. ``rocDecCreateVideoParser()`` returns a handle to the parser. For example:

> 

Reviewer 2: Suggest changing "Parsing parameters" to "The parser parameters", which are the properties of a new parser.

    packet.payload = frames[i].data();
    rocDecParseVideoData(parser_handle, &packet);

Three callbacks must be registered when the parser is used: ``pfn_decode_picture``, ``pfn_sequence_callback``, and ``pfn_display_picture``. These callbacks are triggered in the ``rocDecParseVideoData()`` call.

``pfn_decode_picture`` is triggered when a picture is ready for decoding. Its implementation must call ``rocDecDecodeFrame()`` from the hardware decoder API. 

``pfn_sequence_callback`` is triggered when a new sequence header is encountered or when there's a format change. Its implementation handles reconfiguring the decoder to handle the new frame format. Its implementation must call ``rocDecReconfigureDecoder()`` from the hardware decoder API.

``pfn_display_picture`` is triggered when a frame has been decoded.  Its implementation must call ``rocDecGetVideoFrame()`` from the hardware decoder API. 

A fourth callback, ``pfn_get_sei_msg``, is optional. ``pfn_get_sei_msg`` is triggered when a Supplementation Enhancement Information (SEI) message is parsed and returned to the caller. 

If any of the callbacks return an error, the error is propagated back to the application.

> Once the stream is fully decoded, ``rocDecDestroyVideoParser()`` must be called oto destroy the parser object and free all allocated resources.

Reviewer: ```suggestion
Once the stream is fully decoded, ``rocDecDestroyVideoParser()`` must be called to destroy the parser object and free all allocated resources.

```
